eth: close of closed channel on shutdown

at 82ac4712f17d69e6fa90949357ba0d97a3d8a770. 

```
panic: close of closed channel

goroutine 9 [running]:
github.com/ethereum/go-ethereum/eth.(*ProtocolManager).Stop(0xc208176600)
        /home/staff/fjl/develop/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/eth/handler.go:124 +0x179
github.com/ethereum/go-ethereum/eth.(*Ethereum).Stop(0xc20816a200)
        /home/staff/fjl/develop/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/eth/backend.go:501 +0x66
github.com/ethereum/go-ethereum/cmd/utils.func·002(0x2b8dd32fc100, 0xc21f81e238)
        /home/staff/fjl/develop/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/cmd/utils/cmd.go:121 +0x2a
github.com/ethereum/go-ethereum/cmd/utils.RunInterruptCallbacks(0x2b8dd32fc100, 0xc21f81e238)
        /home/staff/fjl/develop/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/cmd/utils/cmd.go:61 +0x77
github.com/ethereum/go-ethereum/cmd/utils.func·001()
        /home/staff/fjl/develop/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/cmd/utils/cmd.go:54 +0x241
created by github.com/ethereum/go-ethereum/cmd/utils.HandleInterrupt
        /home/staff/fjl/develop/go-ethereum/build/_workspace/src/github.com/ethereum/go-ethereum/cmd/utils/cmd.go:56 +0xb5
```

--- a/cmd/geth/main.go
+++ b/cmd/geth/main.go
@@ -347,7 +347,6 @@ func main() {
 }
 
 func run(ctx *cli.Context) {
-	utils.HandleInterrupt()
 	cfg := utils.MakeEthConfig(ClientIdentifier, nodeNameVersion, ctx)
 	ethereum, err := eth.New(cfg)
 	if err != nil {
@@ -527,10 +526,9 @@ func blockRecovery(ctx *cli.Context) {
 
 func startEth(ctx *cli.Context, eth *eth.Ethereum) {
 	// Start Ethereum itself
-
 	utils.StartEthereum(eth)
-	am := eth.AccountManager()
 
+	am := eth.AccountManager()
 	account := ctx.GlobalString(utils.UnlockedAccountFlag.Name)
 	accounts := strings.Split(account, " ")
 	for i, account := range accounts {

--- a/cmd/utils/cmd.go
+++ b/cmd/utils/cmd.go
@@ -46,29 +46,6 @@ const (
 
 var interruptCallbacks = []func(os.Signal){}
 
-// Register interrupt handlers callbacks
-func RegisterInterrupt(cb func(os.Signal)) {
-	interruptCallbacks = append(interruptCallbacks, cb)
-}
-
-// go routine that call interrupt handlers in order of registering
-func HandleInterrupt() {
-	c := make(chan os.Signal, 1)
-	go func() {
-		signal.Notify(c, os.Interrupt)
-		for sig := range c {
-			glog.V(logger.Error).Infof("Shutting down (%v) ... \n", sig)
-			RunInterruptCallbacks(sig)
-		}
-	}()
-}
-
-func RunInterruptCallbacks(sig os.Signal) {
-	for _, cb := range interruptCallbacks {
-		cb(sig)
-	}
-}
-
 func openLogFile(Datadir string, filename string) *os.File {
 	path := common.AbsolutePath(Datadir, filename)
 	file, err := os.OpenFile(path, os.O_RDWR|os.O_CREATE|os.O_APPEND, 0666)
@@ -149,19 +126,24 @@ func StartEthereum(ethereum *eth.Ethereum) {
 	if err := ethereum.Start(); err != nil {
 		Fatalf("Error starting Ethereum: %v", err)
 	}
-	RegisterInterrupt(func(sig os.Signal) {
+	go func() {
+		sigc := make(chan os.Signal, 1)
+		signal.Notify(sigc, os.Interrupt)
+		defer signal.Stop(sigc)
+		<-sigc
+		glog.V(logger.Info).Infoln("Got interrupt, shutting down...")
 		ethereum.Stop()
 		logger.Flush()
-	})
-}
-
-func StartEthereumForTest(ethereum *eth.Ethereum) {
-	glog.V(logger.Info).Infoln("Starting ", ethereum.Name())
-	ethereum.StartForTest()
-	RegisterInterrupt(func(sig os.Signal) {
-		ethereum.Stop()
-		logger.Flush()
-	})
+		for i := 10; i > 0; i-- {
+			<-sigc
+			if i > 1 {
+				glog.V(logger.Info).Infoln("Already shutting down, please be patient.")
+				glog.V(logger.Info).Infoln("Interrupt", i-1, "more times to induce panic.")
+			}
+		}
+		glog.V(logger.Error).Infof("Force quitting: this might not end so well.")
+		panic("boom")
+	}()
 }
 
 func FormatTransactionData(data string) []byte {

